Objects/unionobject.c

```diff
@@ -0,0 +1,174 @@
+// types.Union -- used to represent e.g. Union[int, str], int | str
+#include "Python.h"
+#include "pycore_object.h"
+#include "structmember.h"
+
+typedef struct {
+    PyObject_HEAD;
+    PyObject *args;
+} unionobject;
+
+static void
+unionobject_dealloc(PyObject *self)
+{
+    unionobject *alias = (unionobject *)self;
+
+    _PyObject_GC_UNTRACK(self);
+    Py_XDECREF(alias->args);
+    self->ob_type->tp_free(self);
+}
+
+
+static Py_hash_t
+union_hash(PyObject *self)
+{
+    unionobject *alias = (unionobject *)self;
+    Py_hash_t h1 = PyObject_Hash(alias->args);
+    if (h1 == -1) {
+        return -1;
+    }
+    return h1;
+}
+
+static int
+union_traverse(PyObject *self, visitproc visit, void *arg)
+{
+    unionobject *alias = (unionobject *)self;
+    Py_VISIT(alias->args);
+    return 0;
+}
+
+static PyMemberDef union_members[] = {
+    {"__args__", T_OBJECT, offsetof(unionobject, args), READONLY},
+    {0}
+};
+
+static PyObject *
+union_getattro(PyObject *self, PyObject *name)
+{
+    return PyObject_GenericGetAttr(self, name);
+}
+
+// TODO: MM: Implement this for isinstance checks
+static PyObject *
+union_instancecheck(PyObject *self, PyObject *instance)
+{
+    unionobject *alias = (unionobject *) self;
+    Py_ssize_t nargs = PyTuple_GET_SIZE(alias->args);
+    int retval;
+    for (Py_ssize_t iarg = 0; iarg < nargs; iarg++) {
+        PyObject *arg = PyTuple_GET_ITEM(alias->args, iarg);
+        if (PyType_Check(arg)) {
+            // MM: TODO: Find out best method to use for this.
+            retval = PyObject_IsInstance(instance, arg);
+        }
+    }
+    return self;
+}
+
+static int
+is_typing_name(PyObject *obj, char *name)
+{
+    PyTypeObject *type = Py_TYPE(obj);
+    if (strcmp(type->tp_name, name) != 0) {
+        return 0;
+    }
+    PyObject *module = PyObject_GetAttrString((PyObject *)type, "__module__");
+    if (module == NULL) {
+        return -1;
+    }
+    int res = PyUnicode_Check(module)
+        && _PyUnicode_EqualToASCIIString(module, "typing");
+    Py_DECREF(module);
+    return res;
+}
+
+static PyMethodDef union_methods[] = {
+    {"__instancecheck__", union_instancecheck, METH_O},
+    {0}
+};
+
+static PyObject *
+union_richcompare(PyObject *a, PyObject *b, int op)
```

Comment:
This should presumably convert each of the args to a set and then compare the sets. (This way, we preserve the *visual* difference between `int|str` and `str|int`, but they still compare equal. Sort of like how two dicts with different item orders but the same sets of items print differently but compare equal.)